File: server/web/api/live_torrent_test.go
package api_test

import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"strings"
	"testing"
	"time"
)

// Live torrent HTTP checks against a running TorrServer (data/ settings, real hash).
//
//	LIVE_TORRENT=1 TORRSERVER_URL=http://127.0.0.1:18091 TS_USER=ts TS_PASS=ts \
//	  go test ./web/api -run TestLiveTorrent -count=1 -timeout 3m
func TestLiveTorrentStreamStatPlayFFP(t *testing.T) {
	base := strings.TrimRight(os.Getenv("TORRSERVER_URL"), "/")
	if os.Getenv("LIVE_TORRENT") != "1" || base == "" {
		t.Skip("set LIVE_TORRENT=1 TORRSERVER_URL=http://127.0.0.1:PORT")
	}
	hash := os.Getenv("LIVE_TORRENT_HASH")
	if hash == "" {
		hash = "e5a5bdb8ff6152657a1e051024b618dd37d76957"
	}
	user := getenv("TS_USER", "ts")
	pass := getenv("TS_PASS", "ts")

	client := &http.Client{Timeout: 60 * time.Second}
	do := func(method, url, body string, extra map[string]string) *http.Response {
		t.Helper()
		var rdr io.Reader
		if body != "" {
			rdr = strings.NewReader(body)
		}
		req, err := http.NewRequest(method, url, rdr)
		if err != nil {
			t.Fatal(err)
		}
		req.SetBasicAuth(user, pass)
		req.Header.Set("Accept", "application/json")
		if body != "" {
			req.Header.Set("Content-Type", "application/json")
		}
		for k, v := range extra {
			req.Header.Set(k, v)
		}
		res, err := client.Do(req)
		if err != nil {
			t.Fatal(err)
		}
		return res
	}

	readClose := func(res *http.Response) []byte {
		t.Helper()
		defer func() { _ = res.Body.Close() }()
		raw, err := io.ReadAll(res.Body)
		if err != nil {
			t.Fatal(err)
		}
		return raw
	}

	echo := do(http.MethodGet, base+"/echo", "", nil)
	b := readClose(echo)
	if echo.StatusCode != 200 || !strings.Contains(string(b), "MatriX") {
		t.Fatalf("echo %d %s", echo.StatusCode, b)
	}

	add := do(http.MethodPost, base+"/torrents", `{"action":"add","link":"`+hash+`","save_to_db":true}`, nil)
	_ = readClose(add)

	var stat map[string]any
	deadline := time.Now().Add(2 * time.Minute)
	for time.Now().Before(deadline) {
		res := do(http.MethodGet, base+"/stream?link="+hash+"&stat", "", nil)
		raw := readClose(res)
		if res.StatusCode == 200 && json.Unmarshal(raw, &stat) == nil {
			if files, _ := stat["file_stats"].([]any); len(files) > 0 {
				break
			}
		}
		stat = nil
		time.Sleep(2 * time.Second)
	}
	if stat == nil {
		t.Fatal("stream?stat never returned file_stats")
	}
	files, _ := stat["file_stats"].([]any)
	fileID := "1"
	if len(files) > 0 {
		if m, ok := files[0].(map[string]any); ok {
			if v, ok := m["id"].(float64); ok {
				fileID = fmt.Sprintf("%d", int(v))
			}
		}
	}

	play := do(http.MethodGet, base+"/stream?link="+hash+"&index="+fileID+"&play", "", map[string]string{
		"Range": "bytes=0-1023",
	})
	chunk := readClose(play)
	if play.StatusCode != 200 && play.StatusCode != 206 {
		n := min(len(chunk), 200)
		t.Fatalf("play %d %s", play.StatusCode, chunk[:n])
	}
	if len(chunk) == 0 {
		t.Fatal("empty play body")
	}

	ffp := do(http.MethodGet, base+"/ffp/"+hash+"/"+fileID, "", nil)
	ffpBody := readClose(ffp)
	if ffp.StatusCode == 200 {
		var probe map[string]any
		if err := json.Unmarshal(ffpBody, &probe); err != nil {
			n := min(len(ffpBody), 200)
			t.Fatalf("ffp json: %v %s", err, ffpBody[:n])
		}
		if probe["format"] == nil && probe["streams"] == nil {
			n := min(len(ffpBody), 200)
			t.Fatalf("ffp missing format/streams: %s", ffpBody[:n])
		}
	} else {
		n := min(len(ffpBody), 200)
		t.Logf("ffp non-200 (%d) — ffprobe may be missing: %s", ffp.StatusCode, ffpBody[:n])
	}
}

func getenv(key, fallback string) string {
	if v := os.Getenv(key); v != "" {
		return v
	}
	return fallback
}

File: web/e2e/live-torrent.test.mjs
/**
 * Live torrent / stream / ffp checks.
 *
 *   LIVE=1 TORRSERVER_URL=http://127.0.0.1:18091 TS_USER=ts TS_PASS=ts yarn test:e2e:live
 *
 * Optional: LIVE_TORRENT_HASH (default e5a5bdb8…), TORZNAB_QUERY.
 * Indexer hosts are taken from the running server settings, or TORZNAB_HOST + TORZNAB_KEY.
 */
import assert from 'node:assert/strict'
import { test } from 'node:test'

import { chromium } from 'playwright'

const BASE = process.env.TORRSERVER_URL || ''
const USER = process.env.TS_USER || 'ts'
const PASS = process.env.TS_PASS || 'ts'
const HASH = (process.env.LIVE_TORRENT_HASH || 'e5a5bdb8ff6152657a1e051024b618dd37d76957').toLowerCase()
const live = process.env.LIVE === '1' && BASE !== ''

function authHeader() {
  return 'Basic ' + Buffer.from(`${USER}:${PASS}`).toString('base64')
}

async function api(path, { method = 'GET', json, headers = {} } = {}) {
  const res = await fetch(BASE + path, {
    method,
    headers: {
      Authorization: authHeader(),
      Accept: 'application/json',
      ...(json ? { 'Content-Type': 'application/json' } : {}),
      ...headers,
    },
    body: json ? JSON.stringify(json) : undefined,
  })
  return res
}

function skipIfNotLive(t) {
  if (!live) t.skip('set LIVE=1 TORRSERVER_URL=http://127.0.0.1:PORT')
}

test('live echo is MatriX', { timeout: 15_000 }, async t => {
  skipIfNotLive(t)
  const res = await api('/echo')
  const body = await res.text()
  assert.equal(res.status, 200)
  assert.match(body, /MatriX/)
})

test('live torrent get/add + stream stat + play range + ffp', { timeout: 180_000 }, async t => {
  skipIfNotLive(t)

  let got = await api('/torrents', { method: 'POST', json: { action: 'get', hash: HASH } })
  if (got.status !== 200) {
    const add = await api('/torrents', {
      method: 'POST',
      json: { action: 'add', link: HASH, save_to_db: true, title: 'live-e2e' },
    })
    assert.ok(add.status === 200, `add torrent failed: ${add.status} ${await add.text()}`)
  }

  let torrent
  const deadline = Date.now() + 120_000
  while (Date.now() < deadline) {
    const statRes = await fetch(`${BASE}/stream?link=${HASH}&stat`, {
      headers: { Authorization: authHeader(), Accept: 'application/json' },
    })
    if (statRes.ok) {
      torrent = await statRes.json()
      const files = torrent.file_stats || torrent.FileStats || []
      if (files.length > 0 && torrent.stat !== 1) break
    }
    await new Promise(r => setTimeout(r, 2000))
  }
  assert.ok(torrent, 'no stream?stat payload')
  const files = torrent.file_stats || []
  assert.ok(files.length > 0, `no file_stats: ${JSON.stringify(torrent).slice(0, 400)}`)
  const fileId = files[0].id ?? files[0].Id ?? 1

  const play = await fetch(`${BASE}/stream?link=${HASH}&index=${fileId}&play`, {
    headers: { Authorization: authHeader(), Range: 'bytes=0-1023' },
  })
  const playStatus = play.status
  const buf = Buffer.from(await play.arrayBuffer())
  assert.ok([200, 206].includes(playStatus), `play status ${playStatus}`)
  assert.ok(buf.length > 0, 'empty play body')

  const ffp = await api(`/ffp/${HASH}/${fileId}`)
  const ffpStatus = ffp.status
  const ffpText = await ffp.text()
  if (ffpStatus === 200) {
    const probe = JSON.parse(ffpText)
    assert.ok(probe.format || probe.streams, `ffp json missing format/streams: ${ffpText.slice(0, 200)}`)
  } else {
    t.diagnostic(`ffp skipped/non-200: ${ffpStatus} ${ffpText.slice(0, 200)}`)
  }
})

test('live torznab search (server settings or TORZNAB_HOST)', { timeout: 60_000 }, async t => {
  skipIfNotLive(t)
  const setsRes = await api('/settings', { method: 'POST', json: { action: 'get' } })
  assert.equal(setsRes.status, 200)
  const sets = await setsRes.json()
  const fromEnv = process.env.TORZNAB_HOST
  const enabled = Boolean(fromEnv) || Boolean(sets.EnableTorznabSearch && (sets.TorznabUrls || []).length)
  if (!enabled) {
    t.skip('no Torznab indexers on server; set TORZNAB_HOST or enable in data/settings.json')
    return
  }
  const q = encodeURIComponent(process.env.TORZNAB_QUERY || 'matrix')
  const search = await api(`/torznab/search/?query=${q}`)
  const searchStatus = search.status
  const xml = await search.text()
  assert.ok(searchStatus === 200, `torznab search ${searchStatus} ${xml.slice(0, 200)}`)
  let parsed = null
  try {
    parsed = JSON.parse(xml)
  } catch {
    /* xml */
  }
  if (Array.isArray(parsed)) {
    t.diagnostic(`torznab results: ${parsed.length}`)
  } else {
    assert.match(xml, /rss|error|item|channel/i)
  }
})

test('live Playwright library card for hash', { timeout: 60_000 }, async t => {
  skipIfNotLive(t)
  const browser = await chromium.launch({ headless: true })
  const context = await browser.newContext({
    locale: 'en-US',
    viewport: { width: 1440, height: 900 },
    extraHTTPHeaders: { 'Accept-Language': 'en-US,en;q=0.9' },
  })
  const page = await context.newPage()
  try {
    await page.goto(BASE + '/', { waitUntil: 'domcontentloaded', timeout: 30_000 })
    await page.waitForSelector('input[name="username"], .torrent-card', { timeout: 20_000 })
    const userBox = page.locator('input[name="username"]')
    if (await userBox.isVisible().catch(() => false)) {
      await userBox.fill(USER)
      await page.locator('input[name="password"]').fill(PASS)
      await page.getByRole('button', { name: /sign in|войти/i }).click()
    }
    await page.waitForFunction(
      h =>
        [...document.querySelectorAll('.torrent-card')].some(
          el => (el.getAttribute('data-hash') || '').toLowerCase() === h,
        ),
      HASH,
      { timeout: 30_000 },
    )
  } finally {
    await context.close()
    await browser.close()
  }
})
